gazebo_ros/src/executor.cpp

```diff
@@ -109,31 +112,38 @@ Executor::run()
 {
   while (rclcpp::ok(this->context_) && spinning.load()) {
     rclcpp::AnyExecutable any_exec;
-    std::lock_guard<std::mutex> wait_lock(wait_mutex_);
-    if (!rclcpp::ok(this->context_) || !spinning.load()) {
-      return;
-    }
+    add_remove_node_high_priority_mutex_.lock();
+    std::lock_guard<std::mutex> add_remove_node_lp_lock(add_remove_node_low_priority_mutex_);
```

Comment:
you have to swap the lines here.
IIUC the idea is that the low priority mutex can only be taken by one low priority thread in the whole critical section.
By taking the hpmutex, then the data mutex, and then unlocking the hpmutex, you notifiy all high priority threads and not the low priority ones (which will then wait in the data mutex).

Reply:
Oops, yeah you're correct.